KEP-4193: Update ServiceAccount admin docs with additional info on bound tokens (#48495)
kubernetes/website: content/en/docs/reference/access-authn-authz/service-accounts-admin.md

* Pod (used for projected volume mounts, see below)
* Secret (can be used to allow revoking a token by deleting the Secret)

> * Node (in v1.31, creating new node-bound tokens is beta, using existing node-bound tokens is GA)

Reviewer: reference https://github.com/kubernetes/kubernetes/blob/8770bd58d04555303a3a15b30c245a58723d0f4a/pkg/features/versioned_kube_features.go#L706 for node binding going GA in 1.32

Hence `kubectl get tokenreview` is not a valid command.
{{< /note >}}

#### Schema for service account private claims

The schema for the Kubernetes-specific claims within JWT tokens is not currently documented,

> however the relevant code area can be found in [the serviceaccount package]() in the Kubernetes codebase.
> 
> [the serviceaccount package]: https://github.com/kubernetes/kubernetes/blob/d8919343526597e0788a1efe133c70d9a0c07f69/pkg/serviceaccount/claims.go#L56-L68

Reviewer 2: Pls use inline links because otherwise our link checker is unable to parse/check it.

```suggestion
however the relevant code area can be found in
[the serviceaccount package](https://github.com/kubernetes/kubernetes/blob/d8919343526597e0788a1efe133c70d9a0c07f69/pkg/serviceaccount/claims.go#L56-L68)
in the Kubernetes codebase.
```

> You can inspect a JWT using standard JWT decoding tools, an example of a JWT for the

Reviewer 2: ```suggestion
You can inspect a JWT using standard JWT decoding tool. Below is an example of a JWT for the
```

`my-serviceaccount` ServiceAccount, bound to a Pod object named `my-pod` which is scheduled
to the Node `my-node`, in the `default` namespace:

```json
{
  "aud": [

>     "https://kubernetes.default.svc.cluster.local"

Reviewer: make it clear with a comment that this is an example and can vary from cluster to cluster, and from token to token (you can switch the formatting type to yaml and add # comments that will get formatted nicely) (or better yet, make this clearly a dummy value like `https://my-audience.example.com`)

  ],
  "exp": 1729605240,
  "iat": 1729601640,

>   "iss": "https://kubernetes.default.svc.cluster.local",

Reviewer: same comment that this is an example and can vary from cluster to cluster (or better yet, make this clearly a dummy value like `https://my-cluster.example.com`)

  "jti": "aed34954-b33a-4142-b1ec-389d6bbb4936",
  "kubernetes.io": {
    "namespace": "default",
    "node": {
      "name": "my-node",
      "uid": "646e7c5e-32d6-4d42-9dbd-e504e6cbe6b1"
    },
    "pod": {
      "name": "my-pod",
      "uid": "5e0bd49b-f040-43b0-99b7-22765a53f7f3"
    },
    "serviceaccount": {
      "name": "my-serviceaccount",
      "uid": "14ee3fa4-a7e2-420f-9f9a-dbc4507c3798"
    }
  },
  "nbf": 1729601640,
  "sub": "system:serviceaccount:default:my-serviceaccount"
}
```

{{< note >}}
The presence of both the `pod` and `node` claim implies that this token is bound

> to a *Pod* object. When verifying Pod bound ServiceAccount tokens, the apiserver **does not**

Reviewer 2: ```suggestion
to a *Pod* object. When verifying Pod bound ServiceAccount tokens, the API server **does not**
```

verify the existence of the referenced Node object.
{{< /note >}}

Services that run outside of Kubernetes and want to perform offline validation of JWTs may
use this schema, along with a compliant JWT validator configured with OpenID Discovery information

> from the kube-apiserver, to verify presented JWTs without requiring use of the TokenReview API.

Reviewer 2: ```suggestion
from the API server, to verify presented JWTs without requiring use of the TokenReview API.
```

Services that verify JWTs in this way **do not verify** the claims embedded in the JWT token to be
current and still valid.
This means if the token is bound to an object, and that object no longer exists, the token will still

> be considered valid (until the configured token expiry).

Reviewer 2: ```suggestion
be considered valid (until the configured token expires).
```

Clients that require assurance that a token's bound claims are still valid **MUST** use the TokenReview
API to present the token to the `kube-apiserver` for it to verify and expand the embedded claims, using
similar steps to the [Verifying and inspecting private claims](#verifying-and-inspecting-private-claims)

> section above, but with a [supported client library](/docs/reference/using-api/client-libraries/).

Author: I'm planning to add an example on using the TokenReview API to verify a claim to the `k8s.io/client-go/examples` package, which we can then link to from here (as well as via the 'Examples' link on this client libraries page).

Author: ... unless you want to avoid a bias towards client-go and leave people to 'discover' the example by clicking through from this link :)

Author: opened https://github.com/kubernetes/kubernetes/pull/128268

Author: Do we want an example of manually inspecting/validating JWTs without TokenReview? It seems sensible for our best-practice documentation to detail how to use the proper tools, but we do also publish OpenID Discovery documents so that users *can* do this. If so, which JWT/OpenID discovery libraries are preferable? A manual HTTP request to fetch the signing information, followed by using whatever JWT handling library the apiserver already uses?

Reviewer: we can link to jwt RFCs if we want

Reviewer: this bit lgtm

Hence `kubectl get tokenreview` is not a valid command.
{{< /note >}}

#### Schema for service account private claims

The schema for the Kubernetes-specific claims within JWT tokens is not currently documented,
however the relevant code area can be found in
[the serviceaccount package](https://github.com/kubernetes/kubernetes/blob/d8919343526597e0788a1efe133c70d9a0c07f69/pkg/serviceaccount/claims.go#L56-L68)
in the Kubernetes codebase.

You can inspect a JWT using standard JWT decoding tool. Below is an example of a JWT for the
`my-serviceaccount` ServiceAccount, bound to a Pod object named `my-pod` which is scheduled

> to the Node `my-node`, in the `default` namespace:

Reviewer: ```suggestion
to the Node `my-node`, in the `my-namespace` namespace:
```

```json
{
  "aud": [
    "https://my-audience.example.com"
  ],
  "exp": 1729605240,
  "iat": 1729601640,
  "iss": "https://my-cluster.example.com",
  "jti": "aed34954-b33a-4142-b1ec-389d6bbb4936",
  "kubernetes.io": {

>     "namespace": "default",

Reviewer: ```suggestion
    "namespace": "my-namespace",
```

    "node": {
      "name": "my-node",
      "uid": "646e7c5e-32d6-4d42-9dbd-e504e6cbe6b1"
    },
    "pod": {
      "name": "my-pod",
      "uid": "5e0bd49b-f040-43b0-99b7-22765a53f7f3"
    },
    "serviceaccount": {
      "name": "my-serviceaccount",
      "uid": "14ee3fa4-a7e2-420f-9f9a-dbc4507c3798"
    }
  },
  "nbf": 1729601640,

>   "sub": "system:serviceaccount:default:my-serviceaccount"

Reviewer: ```suggestion
  "sub": "system:serviceaccount:my-namespace:my-serviceaccount"
```
